File: 백준/Gold/2665. 미로만들기/미로만들기.py
import sys
from collections import deque

input = sys.stdin.readline

dx = [0, 0, 1, -1]
dy = [1, -1, 0, 0]

def bfs():
    queue = deque()
    queue.append((0, 0))

    visit[0][0] = 0

    while queue:
        x, y = queue.popleft()

        if x == n - 1 and y == n - 1:
            return visit[x][y]
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0 <= nx < n and 0 <= ny < n and visit[nx][ny] == -1:
                if maps[nx][ny]:
                    visit[nx][ny] = visit[x][y]
                    queue.appendleft((nx, ny))
                else:
                    visit[nx][ny] = visit[x][y] + 1
                    queue.append((nx, ny))

n = int(input())
maps = [list(map(int, input().rstrip())) for _ in range(n)]
visit = [[-1] * n for _ in range(n)]
print(bfs())

File: 백준/Gold/2665. 미로만들기/미로만들기.java
import java.io.*;
import java.util.*;

class Main {
    static int n;
    static int [][] maps;
    static int [][] visit;
    static int [] dx = {0, 0, 1, -1};
    static int [] dy = {1, -1, 0, 0};
    public static class Point {
        int x, y;
        public Point(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
    private static void BFS(Point start) {
        visit[start.x][start.y] = 1;
        Deque<Point> queue = new LinkedList<Point>();
        queue.offer(start);

        while (!queue.isEmpty()) {
            int size = queue.size();

            for (int s = 0; s < size; s++) {
                Point current = queue.poll();

                for (int i = 0; i < 4; i++) {
                    int nx = current.x + dx[i];
                    int ny = current.y + dy[i];

                    if (0 <= nx && nx < n && 0 <= ny && ny < n && visit[nx][ny] == 0) {
                        if (maps[nx][ny] == 1) {
                            queue.offerFirst(new Point(nx, ny));
                            visit[nx][ny] = visit[current.x][current.y];
                        }
                        else {
                            queue.offer(new Point(nx, ny));
                            visit[nx][ny] = visit[current.x][current.y] + 1;
                        }
                    }
                }
            }
        }

    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        n = Integer.parseInt(br.readLine());
        visit = new int [n][n];
        maps = new int [n][n];
        
        for (int i = 0; i < n; i++) {
            String [] row = br.readLine().split("");
            for (int j = 0; j < n; j++) {
                maps[i][j] = Integer.parseInt(row[j]);
            }
        }

        BFS(new Point(0, 0));
        System.out.println(visit[n - 1][n - 1] - 1);
    }
}